### PR TITLE
Fix 5592: Colon (:) in in object_store::path::{Path} is not handled on Windows

### LINKED ISSUE #5592: Colon (:) in in object_store::path::{Path} is not handled on Windows
**Describe the bug**
When creating a `Path` on Windows colon (`:`) is not sanitized like `<` and `|`.

**To Reproduce**
```rust
use object_store::path::{Path};

fn main() {
    let path: String = r"C:\table\time=2021-01-02 03:04:06.000003\file.parquet".to_string();
    // let path: Result<Path, object_store::path::Error> = Path::parse(path);
    let path_from = Path::from(path);
    println!( "{:?}", path_from);

    let path: String = r"C:\table\time=2021-01-02 03:04:06.000003\file.parquet".to_string();
    let path_parse: Result<Path, object_store::path::Error> = Path::parse(path);
    println!( "{:?}", path_parse);

    let path: String = r"C:\table\time=2021-01-02 03:04:06.000003\<file|.parquet".to_string();
    // let path: Result<Path, object_store::path::Error> = Path::parse(path);
    let path_from = Path::from(path);
    println!( "{:?}", path_from);

    let path: String = r"C:\table\time=2021-01-02 03:04:06.000003\<file|.parquet".to_string();
    let path_parse: Result<Path, object_store::path::Error> = Path::parse(path);
    println!( "{:?}", path_parse);
}

```
```
Path { raw: "C:%5Ctable%5Ctime=2021-01-02 03:04:06.000003%5Cfile.parquet" }
Ok(Path { raw: "C:\\table\\time=2021-01-02 03:04:06.000003\\file.parquet" })
Path { raw: "C:%5Ctable%5Ctime=2021-01-02 03:04:06.000003%5C%3Cfile%7C.parquet" }
Ok(Path { raw: "C:\\table\\time=2021-01-02 03:04:06.000003\\<file|.parquet" })
```

See SO for forbidden chars https://stackoverflow.com/questions/1976007/what-characters-are-forbidden-in-windows-and-linux-directory-names
**Expected behavior**
I would have expected the output to be
```
Path { raw: "C:%5Ctable%5Ctime=2021-01-02 03%3A04%3A06.000003%5Cfile.parquet" }
Ok(Path { raw: "C:\\table\\time=2021-01-02 03:04:06.000003\\file.parquet" })
Path { raw: "C:%5Ctable%5Ctime=2021-01-02 03%3A04%3A06.000003%5C%3Cfile%7C.parquet" }
Ok(Path { raw: "C:\\table\\time=2021-01-02 03:04:06.000003\\<file|.parquet" })
```
**Additional context**
This came up here https://github.com/delta-io/delta-rs/issues/2382

### PR DESCRIPTION
**Pull request by hesampakdaman**
# Which issue does this PR close?

Closes #5592.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by thomasfrederikhoeck** [comment]
@hesampakdaman is this ready for review? :-)

**Comment by hesampakdaman** [comment]
@thomasfrederikhoeck I believe so, although I haven't been able to run an integration (don't have Windows accessible).

**Comment by alamb** [comment]
Can someone please verify that this works on Windows? I don't have access to a windows machine at the moment

**Comment by thomasfrederikhoeck** [comment]
I don't think this fixes it. If I run the following on Windows with the fix
```rust
use object_store::path::{Path};

fn main() {
    let path: String = r"C:\table\time=2021-01-02 03:04:06.000003\file.parquet".to_string();
    let path_from: Path = Path::from(path);
    println!( "{:?}", path_from);

    let path: String = r"C:\table\time=2021-01-02 03:04:06.000003\file.parquet".to_string();
    let path_parse: Result<Path, object_store::path::Error> = Path::parse(path);
    println!( "{:?}", path_parse);
}
```

I still see:
```rust
Path { raw: "C:%5Ctable%5Ctime=2021-01-02 03:04:06.000003%5Cfile.parquet" }
Ok(Path { raw: "C:\\table\\time=2021-01-02 03:04:06.000003\\file.parquet" })
```
where I would have expected:
```rust
Path { raw: "C:%5Ctable%5Ctime=2021-01-02 03%3A04%3A06.000003%5Cfile.parquet" }
Ok(Path { raw: "C:\\table\\time=2021-01-02 03:04:06.000003\\file.parquet" })
```
But I might be assuming something wrong? @alamb @hesampakdaman

**Comment by hesampakdaman** [comment]
@thomasfrederikhoeck You're not thinking wrong, my initial suggestion was to change how `Path` behaves. In the discussions we had in the issue, @tustvold mentioned that that would be a breaking change and affect all paths for all stores.

So instead we agreed to change how `LocalFileSystem` behaves when using it from Windows.  If you use the put/get methods it should convert the `:` to `%3A`. Although, I'm a little uncertain if I'm correctly handling the leading `C:\`.

At any rate, if you could test this that would be awesome! I would have done it myself but I don't have access to Windows unfortunately.

**Comment by thomasfrederikhoeck** [comment]
@hesampakdaman I just tested with below and it doesn't appear like it is handeling the `:` in the drive correctly which also are changes but shouldn't. This results in the file being created in a relative position with the name as `C:\projects\delta\pathtester\C%3A\Users\tfh\AppData\Local\Temp\.tmpwqJ9tp\file%3Aname.parquet`

image

The script I ran was

```rust
use object_store::path::{Path};
use tempfile::{TempDir};
use object_store::local::LocalFileSystem;
use object_store::ObjectStore;
use bytes::Bytes;

#[tokio::main]
async fn main() {
    let root = TempDir::new().unwrap();
    let integration = LocalFileSystem::new_with_prefix(root.path()).unwrap();
    let path = Path::parse("file%3Aname.parquet").unwrap();
    let location = Path::parse("file:name.parquet").unwrap();

    integration.put(&location, "test".into()).await.unwrap();

    let result = integration
        .get(&location)
        .await
        .unwrap()
        .bytes()
        .await
        .unwrap();
    assert_eq!(result, Bytes::from("test"));
}
```

**Comment by thomasfrederikhoeck** [comment]
I can also add that if I run the following it creates a file at `C:\projects\delta\pathtester\C%3A\file%3Aname.parquet`

```rust
use object_store::path::{Path};
use object_store::local::LocalFileSystem;
use object_store::ObjectStore;
use bytes::Bytes;

#[tokio::main]
async fn main() {
    let integration = LocalFileSystem::new();
    let location = Path::parse("C:\\file:name.parquet").unwrap();
    println!("{}",location);
    integration.put(&location, "test".into()).await.unwrap();

    let result = integration
        .get(&location)
        .await
        .unwrap()
        .bytes()
        .await
        .unwrap();
    assert_eq!(result, Bytes::from("test"));
}
```

**Comment by thomasfrederikhoeck** [comment]
FYI the code I run is running with `C:\projects\delta\pathtester` as cwd :-)

**Comment by hesampakdaman** [comment]
@thomasfrederikhoeck Thank you for testing, it helps a lot! Yes, as suspected I did not handle the leading colon from the drive. I made a change and hopefully it will work now. 

@alamb I marked this as draft until it is tested again by @thomasfrederikhoeck.

**Comment by thomasfrederikhoeck** [comment]
@hesampakdaman I can confirm it gives the expected results for these two :-D :-D :

```python
use object_store::path::{Path};
use object_store::local::LocalFileSystem;
use object_store::ObjectStore;
use bytes::Bytes;
use tempfile::{NamedTempFile, TempDir};

#[tokio::main]
async fn main() {
    let root_temp = TempDir::new().unwrap();

    // full path
    let integration_full = LocalFileSystem::new();
    let location_full = Path::parse(r"C:\Users\tfh\file:name-full.parquet").unwrap();
    println!("{}",location_full);
    integration_full.put(&location_full, "test".into()).await.unwrap();

    let result = integration_full
        .get(&location_full)
        .await
        .unwrap()
        .bytes()
        .await
        .unwrap();
    assert_eq!(result, Bytes::from("test"));


    let integration_prefefix = LocalFileSystem::new_with_prefix(r"C:\Users\tfh\").unwrap();
    let location_prefefix = Path::parse(r"file:name-prefix.parquet").unwrap();
    println!("{}",location_prefefix);
    integration_prefefix.put(&location_prefefix, "test".into()).await.unwrap();
    let result = integration_prefefix
        .get(&location_prefefix)
        .await
        .unwrap()
        .bytes()
        .await
        .unwrap();
    assert_eq!(result, Bytes::from("test"));

}
```

**Comment by hesampakdaman** [comment]
@alamb I'll take a look later in the week and will add it in this PR!

**Comment by hesampakdaman** [comment]
@alamb I finally added a CI job. However, I noticed that some tests failed if I ran them all under windows. Some of these were outside of `local.rs`.  I did two things to address (ignore) this:

1. I run only the tests in local.rs
2. I added a target = "unix" on two tests that were failing under windows.

I will try to verify if the tests are failing without my changes.

**Comment by hesampakdaman** [comment]
[I re-ran the CI tests for windows](https://github.com/apache/arrow-rs/actions/runs/9758826869/job/26934229459?pr=5830) and removed my changes. It seems that they just fail under windows.

File: .github/workflows/object_store.yml [hunk 1]
```diff
@@ -198,3 +198,16 @@ jobs:
         run: cargo build --target wasm32-unknown-unknown
       - name: Build wasm32-wasi
         run: cargo build --target wasm32-wasi
+
+  windows:
```
Comment (alamb) [inline .github/workflows/object_store.yml:202]: 👍

**Review by alamb — APPROVED** [review]
Thank you again  @hesampakdaman and @thomasfrederikhoeck  for this contribution -- this looks good to me. I also appreciate the effort to make a test. 

It looks like this PR has a conflict to resolve but then it should be good to gol

**Comment by hesampakdaman** [comment]
@alamb Done :rocket:

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again everyone!

**Pull request merged**